include/revng/Pipeline/Step.h

```diff
@@ -0,0 +1,140 @@
+/// will contain the element used for perform the computations.
+class Step {
+private:
+  std::string Name;
+  ContainerSet Containers;
+  std::vector<PipeWrapper> Pipes;
+
+public:
+  /// Clones the Targets from the backing containers of this step
+  /// and excutes all the pipes in sequence contained by this step
+  /// and returns the transformed containers.
+  ///
+  /// The contained values stays unchanged.
+  ContainerSet cloneAndRun(Context &Ctx,
+                           const ContainersStatus &Targets,
+                           llvm::raw_ostream *OS = nullptr);
+
+  /// Returns the set of goals that are already contained in the backing
+  /// containers of this step, futhermore adds to the container ToLoad those
+  /// that were not present.
+  ContainersStatus satisfiableGoals(const ContainersStatus &RequiredGoals,
+                                    ContainersStatus &ToLoad) const;
+
+  /// Returns the predicted state of the Input containers status after the
+  /// execution of all the pipes in this step
+  ContainersStatus deduceResults(ContainersStatus Input) const;
+
+  /// Consumes the provided container and merges it into the current one
+  void mergeBack(ContainerSet RunningContainers) {
+    Containers.mergeBack(std::move(RunningContainers));
+  }
+
+  const ContainerSet &getContainers() const { return Containers; }
+
+  ContainerSet &getContainers() { return Containers; }
+
+  void addPipe(PipeWrapper Wrapper) { Pipes.push_back(std::move(Wrapper)); }
+
+  template<typename... PipeWrapperTypes>
+  Step(std::string Name,
+       ContainerSet Containers,
+       PipeWrapperTypes &&...PipeWrappers) :
+    Name(std::move(Name)),
+    Containers(std::move(Containers)),
+    Pipes({ std::forward<PipeWrapperTypes>(PipeWrappers)... }) {}
+
+  llvm::StringRef getName() const { return Name; }
+
+  template<typename OStream>
+  void dump(OStream &OS, size_t Indents = 0) const {
+    indent(OS, Indents);
+    OS << "Step " << Name << ":\n";
+
+    indent(OS, Indents + 1);
+    OS << "Pipes: \n";
+    for (const auto &Pipe : Pipes)
+      Pipe.dump(OS, Indents + 2);
+
+    indent(OS, Indents + 1);
+    OS << " containers: \n";
+    Containers.dump(OS, Indents + 2);
+  }
+
+  /// Drops from the backing containers all the targets presents in containers
+  /// status
+  llvm::Error invalidate(const ContainersStatus &ToRemove);
+
+  void dump() const debug_function { dump(dbg); }
+
+  llvm::Error storeToDisk(llvm::StringRef DirPath) const;
+  llvm::Error loadFromDisk(llvm::StringRef DirPath);
+
+  const ContainerBase *getContainer(llvm::StringRef ContainerName) const {
+    if (not Containers.contains(ContainerName))
+      return nullptr;
+    return &Containers.at(ContainerName);
+  }
+
+  ContainerBase *getOrCreateContainer(llvm::StringRef ContainerName) {
+    if (not Containers.containsOrCanCreate(ContainerName))
+      return nullptr;
+    return &Containers[ContainerName];
+  }
```

Comment:
Given that we're exposing `getContainers` these seem superfluous.
Can we drop these xor `getContainers`?